### PR TITLE
proposed fix for failing tests in MasterRegistrationTest.

### PR DESCRIPTION
**Pull request by rjcausarano**
Waits for all signals to be delivered before shutting down listeners in class DefaultPublisher.

### CODE REVIEW

**Comment by rjcausarano** [comment]
Hey @jubeira, this might be what we were looking for. I went ahead and merged the previous PR that you reviewed and aproved so I don't understand why that commit is appearing as a change in this PR?

Also I'm pretty sure that there's a better way of handling the caught exception in the last commit. I'll make sure to fix that after your review and squash if necessary.

**Comment by rjcausarano** [comment]
Alright this last commit does the following:

- log.error() in catch block
- shutdown latch uses passed timeout
- adds latch.countdown() to onMasterUnregistrationFailure()
- adds new testUnregisterPublisherFailure() test in class MasterRegistrationTest

File: rosjava/src/main/java/org/ros/internal/node/topic/DefaultPublisher.java [hunk 1]
```diff
@@ -107,9 +109,16 @@ public boolean getLatchMode() {
...
+    try{
+      shutdownLatch = new CountDownLatch(listeners.size());
+      signalOnShutdown(timeout, unit);
+      shutdownLatch.await(timeout, unit);
+      outgoingMessageQueue.shutdown();
+      listeners.shutdown();
+    }
+    catch (InterruptedException e) {
+      log.error(e.getMessage(), e);
+    }
```
Comment (jubeira) [inline rosjava/src/main/java/org/ros/internal/node/topic/DefaultPublisher.java:123]: Please shrink the `try / catch` block only to the `await` call, and add a comment saying that in the normal workflow, `signalOnMasterUnregistration**` should finish execution before shutting down the listeners.

File: rosjava/src/main/java/org/ros/internal/node/topic/DefaultPublisher.java [hunk 2]
```diff
@@ -107,9 +109,16 @@ public boolean getLatchMode() {
 
   @Override
   public void shutdown(long timeout, TimeUnit unit) {
-    signalOnShutdown(timeout, unit);
-    outgoingMessageQueue.shutdown();
-    listeners.shutdown();
+    try{
```
Comment (jubeira) [inline rosjava/src/main/java/org/ros/internal/node/topic/DefaultPublisher.java]: nit: space after `try`

File: rosjava/src/test/java/org/ros/internal/node/MasterRegistrationTest.java [hunk 3]
```diff
@@ -73,4 +73,25 @@ public void onStart(ConnectedNode connectedNode) {
...
+      public void onStart(ConnectedNode connectedNode) {
+        publisher = connectedNode.newPublisher("topic", std_msgs.String._TYPE);
+        publisher.addListener(publisherListener);
+      }
+    }, nodeConfiguration);
+    assertTrue(publisherListener.awaitMasterRegistrationSuccess(1, TimeUnit.SECONDS));
+    rosCore.shutdown();
+    publisher.shutdown();
+    assertTrue(publisherListener.awaitMasterUnregistrationFailure(6, TimeUnit.SECONDS));
+  }
```
Comment (jubeira) [inline rosjava/src/test/java/org/ros/internal/node/MasterRegistrationTest.java:96]: :+1: Great!
It's a bit unfortunate that this test will take 10 secs (5 to shutdown the publisher and 5 to shutdown the node), but it's still a good idea to have it.

File: rosjava/src/main/java/org/ros/internal/node/topic/DefaultPublisher.java [hunk 4]
```diff
@@ -107,9 +109,16 @@ public boolean getLatchMode() {
...
+      shutdownLatch = new CountDownLatch(listeners.size());
+      signalOnShutdown(timeout, unit);
+      shutdownLatch.await(timeout, unit);
+      outgoingMessageQueue.shutdown();
+      listeners.shutdown();
+    }
+    catch (InterruptedException e) {
```
Comment (jubeira) [inline rosjava/src/main/java/org/ros/internal/node/topic/DefaultPublisher.java]: nit: please put `catch` statement at the same line of the closing brace above: ` } catch (InterruptedException e) {`

**Review by jubeira — CHANGES_REQUESTED** [review]
Great job, thanks @rjcausarano!
I've left a couple of minor comments; please address them and squash everything into one commit before sending the PR upstream.

**Review by jubeira — APPROVED** [review]
Awesome; LGTM! I think this can be sent upstream like this.

**Pull request merged**